### PR TITLE
implementing removal/addition of parties

### PR DESCRIPTION
**Pull request by tudorcebere**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -356,7 +352,69 @@ mod tests {
         simulation.run().unwrap()
     }
 
-    fn simulate_dkr(keys: &mut Vec<LocalKey>) {
+    fn simulate_dkr_removal(keys: &mut Vec<LocalKey>, remove_party_indexes: Vec<usize>) {
```
Comment (omershlo) [inline src/lib.rs:355]: indexes => indices

File: src/lib.rs [hunk 2]
```diff
@@ -344,6 +327,19 @@ mod tests {
         simulate_signing(offline_sign, b"ZenGo");
     }
 
+    #[test]
+    fn test_remove_sign_rotate_sign() {
```
Comment (omershlo) [inline src/lib.rs:331]: I suggest to add here a line for description (or instead in the readme file) saying that  removing a party is simply done by not broadcasting encrypted messages to this party. If enough parties decided not to broadcast for a specific party index - this party will not be able to reconstruct a rotated key

File: src/lib.rs [hunk 3]
```diff
@@ -356,7 +352,69 @@ mod tests {
...
+                let mut new_remove_party_indexes = remove_party_indexes.clone();
+                new_remove_party_indexes.retain(|value| *value != refresh_message.party_index);
+                refresh_message.remove_party_indexes = new_remove_party_indexes;
+            }
+
+            for (party_index, refresh_bucket) in broadcast_messages.iter_mut() {
+                if refresh_message.remove_party_indexes.contains(party_index) {
+                    continue;
+                }
+                refresh_bucket.push(refresh_message.clone());
```
Comment (omershlo) [inline src/lib.rs:385]: where referesh_bucket is used ?

File: src/lib.rs [hunk 4]
```diff
@@ -356,7 +352,69 @@ mod tests {
...
+            }
+        }
+
+        for remove_party_index in remove_party_indexes.iter() {
+            assert_eq!(broadcast_messages[remove_party_index].len(), 1);
```
Comment (omershlo) [inline src/lib.rs:390]: what is being checked in this assertion ?

File: src/lib.rs (hunk 3 above)
Comment (tudorcebere) [reply src/lib.rs:385]: Each bucket represents the list of messages that a party identified by the key of the bucket will receive (it's a mutable reference to a list: Imagine something like:

1 -> [m1, m2 m3, m4, m5]
2 -> [m1, m2, m3, m4, m5]
3 -> [m1, m2, m3, m4, m5]
4 -> [m1, m2, m3, m4, m5]
5 -> [m1, m2, m3, m4, m5]

If all the parties are not removing anyone. If they want to remove 4, they will be like
1 -> [m1, m2 m3, m4, m5]
2 -> [m1, m2, m3, m4, m5]
3 -> [m1, m2, m3, m4, m5]
4 -> [m4] // it has access only to it's own refresh message
5 -> [m1, m2, m3, m4, m5]

File: src/lib.rs (hunk 4 above)
Comment (tudorcebere) [reply src/lib.rs:390]: In this test, everyone agrees to delete a party (or even multiple ones). I check that the only message that the party has access to it's their own (by default).

File: src/lib.rs (hunk 1 above)
Comment (tudorcebere) [reply src/lib.rs:355]: 👍🏼

File: src/lib.rs (hunk 2 above)
Comment (tudorcebere) [reply src/lib.rs:331]: 👍🏼

File: src/lib.rs [hunk 5]
```diff
@@ -356,7 +352,72 @@ mod tests {
...
+        }
+
+        // keys will be updated to refreshed values
+        for (party, key) in party_key.iter_mut() {
+            if remove_party_indices.contains(party) {
+                continue;
+            }
+
+            RefreshMessage::collect(
+                broadcast_messages[party].clone().as_slice(),
```
Comment (tudorcebere) [inline src/lib.rs:403]: The buckets are used here

File: src/lib.rs [hunk 6]
```diff
@@ -356,7 +352,72 @@ mod tests {
...
+                key,
+                new_dks[party].clone(),
+            )
+            .expect("");
+        }
+
+        for remove_party_index in remove_party_indices {
+            let result = RefreshMessage::collect(
+                &broadcast_messages[&remove_party_index],
```
Comment (tudorcebere) [inline src/lib.rs:412]: The buckets are used here

**Pull request merged**